Keep but deprecate registry v2 schema1 logic and revert to libtrust-key-based engine ID

tiborvass opened the pull request:
Partially reverts #37874

This partially reverts 98fc091.

Because registry v2 schema1 was not officially deprecated and
registries are still relying on it, this patch puts its logic back.

However, registry v1 relics are not added back since v1 logic has been
removed a while ago.

This also fixes an engine upgrade issue in a swarm cluster. It was relying
on the Engine ID to be the same upon upgrade, but the mentioned commit
modified the logic to use UUID and from a different file.

Since the libtrust key is always needed to support v2 schema1 pushes,
that the old engine ID is based on the libtrust key, and that the engine ID
needs to be conserved across upgrades, adding a UUID-based engine ID logic
seems to add more complexity than it solves the problems.

For that reason, this patch also reverts the engine ID changes.

-----

This reverts commit 13b7d11.

-----
Added a test to guard the daemon ID when upgrading.

----
This also adds a deprecation message to the user as well as in the logs.

Commit:
Add deprecation message for schema1

This will add a warning log in the daemon, and will send the message
to be displayed by the CLI.

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
use gotest.tools assertions in docker_cli_push_test.go

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
Add test for keeping same daemon ID on upgrade

Signed-off-by: Tibor Vass <tibor@docker.com>

tiborvass commented on distribution/push_v2.go:
```diff
@@ -180,8 +182,29 @@ func (p *v2Pusher) pushV2Tag(ctx context.Context, ref reference.NamedTagged, id
+		logrus.Warnf("failed to upload schema2 manifest: %v - falling back to schema1", err)
+
+		logrus.Warnf("[DEPRECATION NOTICE] schema1 support will be removed in an upcoming release. Please contact admins of the %s registry NOW to avoid future disruption. Waiting 13 annoying seconds...", reference.Domain(ref))
+		time.Sleep(13 * time.Second)
```
to be debated

thaJeztah replied:
We could randomize it?

If this needs more discussion, might be good to move it to a separate PR to unblock this one

Commit:
validate: temporarily disable deprecate-integration-cli as part of a revert

Signed-off-by: Tibor Vass <tibor@docker.com>

thaJeztah commented on hack/validate/default:
```diff
@@ -14,4 +14,4 @@ export SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 . ${SCRIPTDIR}/changelog-well-formed
 . ${SCRIPTDIR}/changelog-date-descending
-. ${SCRIPTDIR}/deprecate-integration-cli
+#. ${SCRIPTDIR}/deprecate-integration-cli
```
Guess this was to make CI pass?

tiborvass replied:
Yes

thaJeztah commented on daemon/daemon.go:
```diff
@@ -973,15 +974,17 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
+	trustKey, err := libtrust.LoadKeyFile(config.TrustKeyPath)
+	switch {
+	case err == nil:
+		d.ID = trustKey.PublicKey().KeyID()
```
Can we write the ID to the new file in this case? That way we're able to preserve the ID once we really remove this logic.

FWIW, the format of the ID is not formalized (the new code uses an UUID for convenience), so storing the old ID format in the new location should not be a problem

tiborvass replied:
I'll update

tiborvass replied:
@thaJeztah @justincormack the only problem I see here is that we would put a non-UUID into a file called `engine_uuid`. Should we rename the file to `engine_id` ?

thaJeztah replied:
Makes sense to me; it's used to store an ID (and we describe it in the API that the format is not specified)

Commit:
Revert "Remove Schema1 integration test suite"

This reverts commit 13b7d11be1a97ebe82b3a0f0b5db9b873a481637.

Signed-off-by: Tibor Vass <tibor@docker.com>

tiborvass commented:
@thaJeztah @tonistiigi @justincormack PTAL I updated the code, the title, and the description based on the discussions. Removed the annoying sleep, will PR it separately.

Commit:
Revert "Remove the rest of v1 manifest support"

This reverts commit 98fc09128b3f8221680bd22c2de5b893bd3d7ef8 in order to
keep registry v2 schema1 handling and libtrust-key-based engine ID.

Because registry v2 schema1 was not officially deprecated and
registries are still relying on it, this patch puts its logic back.

However, registry v1 relics are not added back since v1 logic has been
removed a while ago.

This also fixes an engine upgrade issue in a swarm cluster. It was relying
on the Engine ID to be the same upon upgrade, but the mentioned commit
modified the logic to use UUID and from a different file.

Since the libtrust key is always needed to support v2 schema1 pushes,
that the old engine ID is based on the libtrust key, and that the engine ID
needs to be conserved across upgrades, adding a UUID-based engine ID logic
seems to add more complexity than it solves the problems.

Hence reverting the engine ID changes as well.

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
Remove v1 manifest code

Signed-off-by: Tibor Vass <tibor@docker.com>

tiborvass commented:
I updated the code by actually reverting the original commit 98fc09128b3f8221680bd22c2de5b893bd3d7ef8 and added a commit that is the change that it should have been. This way it is easier to compare with 98fc09128b3f8221680bd22c2de5b893bd3d7ef8.

Also made a change to send the deprecation message to the client in addition to logging a warning to the daemon.

thaJeztah commented on daemon/images/service.go:
```diff
@@ -14,6 +14,7 @@ import (
 	"github.com/docker/docker/layer"
 	dockerreference "github.com/docker/docker/reference"
 	"github.com/docker/docker/registry"
+	"github.com/docker/libtrust"
```
FWIW; had a chat with @dmcgowan, and there's a desire to get rid of this dependency; not sure if it's worth doing so as part of this PR (if we plan to remove it in the next release); his suggestion was to copy the parts we use into this repository (somewhere in `pkg` or `internal` perhaps?), so that we can remove the dependency

tiborvass replied:
@thaJeztah Sure I can do that in a followup, don't want to add more changes to this PR which is already big.

cpuguy83 reviewed (approved):
LGTM :(

andrewhsu commented:
SGTM

As much as I would prefer quay.io to support v2 schema2 pushes, at this point it fails to handle it. I just tried and daemon says:
```
Jun 18 17:58:29 ip-172-31-27-101 dockerd[1129]: time="2019-06-18T17:58:29.365129917Z" level=warning msg="failed to upload schema2 manifest: manifest invalid: manifest invalid"
```

tiborvass commented:
Thanks for the reviews.

dmcgowan commented:
This message should not have been added on pull, that can be unnecessarily confusing for users as there is nothing that must be done now related to pull and just causes noise. If a user is pushing to a registry which still only supports v1, then they do need to contact their registry administrator to upgrade. https://github.com/docker/distribution/issues/2976
